src/lib.rs

```diff
@@ -19,25 +19,171 @@
+//! ```
+//!
+//! BSON is the primary data represenation for [MongoDB](https://www.mongodb.com/), and this crate is used in the
+//! [`mongodb`](https://docs.rs/mongodb/0.10.0/mongodb/) driver crate in its API and implementation.
+//!
+//! For more information about BSON itself, see [bsonspec.org](http://bsonspec.org).
+//!
+//! ## BSON values
+//!
+//! Many different types can be represented as a BSON value, including 32-bit and 64-bit signed
+//! integers, 64 bit floating point numbers, strings, datetimes, embedded documents, and more. To
+//! see a full list of possible BSON values, see the [BSON specification](http://bsonspec.org/spec.html). The various
+//! possible BSON values are modeled in this crate by the [`Bson`](enum.Bson.html) enum.
+//!
+//! ### Creating [`Bson`](enum.Bson.html) instances
+//!
+//! [`Bson`](enum.Bson.html) values can be instantiated directly or via the
+//! [`bson!`](macro.bson.html) macro:
+//!
+//! ```rust
+//! # use bson::{bson, Bson};
+//! let string = Bson::String("hello world".to_string());
+//! let int = Bson::I32(5);
+//! let array = Bson::Array(vec![Bson::I32(5), Bson::Boolean(false)]);
+//!
+//! let string: Bson = "hello world".into();
+//! let int: Bson = (5 as i32).into();
+//!
+//! let string: Bson = bson!("hello world");
+//! let int: Bson = bson!(5);
+//! let array = bson!([5, false]);
+//! ```
+//! [`bson!`](macro.bson.html) has supports both array and object literals, and it automatically
+//! converts any values specified to [`Bson`](enum.Bson.html), provided they are either `Serialize`
+//! or `Into<Bson>`.
+//!
+//! ### [`Bson`](enum.Bson.html) value unwrapping
+//!
+//! [`Bson`](enum.Bson.html) has a number of helper methods for accessing the underlying native Rust
+//! types. These helpers can be useful in cirumstances in which a the specific type of a BSON value
+//! is known ahead of time.
+//!
+//! e.g.:
+//! ```rust
+//! # use bson::{bson, Bson};
+//! let value = Bson::I32(5);
+//! let int = value.as_i32(); // Some(5)
+//! let bool = value.as_bool(); // None
+//!
+//! let value = bson!([true]);
+//! let array = value.as_array(); // Some(&Vec<Bson>)
+//! ```
+//!
+//! ## BSON documents
+//!
+//! BSON documents are ordered maps of UTF-8 encoded strings to BSON values. They are logically
+//! similar to JSON objects in that they can contain subdocuments, arrays, and values of several
+//! different types. This crate models BSON documents via the
+//! [`Document`](document/struct.Document.html) struct.
+//!
+//! ### Creating [`Document`](document/struct.Document.html)s
+//!
+//! [`Document`](document/struct.Document.html)s can be created directly either from a byte
+//! reader containing BSON data or via the `doc!` macro:
+//! ```rust
+//! # use bson::doc;
+//! # use std::io::Read;
+//! let mut bytes = hex::decode("0C0000001069000100000000").unwrap();
+//! let doc = bson::decode_document(&mut bytes.as_slice()).unwrap(); // { "i": 1 }
```

Comment:
Not sure if you'd want to fix this in this PR, but this should change to `Document::decode` since we're updating the name [here](https://github.com/mongodb/bson-rust/pull/175#pullrequestreview-418789367)

Reply:
Once your PR gets merged, I'll update this to use the correct method. Until then I'll leave it as-is so the doc tests don't fail.

Reply:
just merged

Reply:
done